statistics/cmsketch.go

```diff
@@ -44,29 +60,173 @@ func NewCMSketch(d, w int32) *CMSketch {
+	lastVal       uint64
+}
+
+func newTopNHelper(sample [][]byte, numTop uint32) *topNHelper {
+	counter := make(map[hack.MutableString]uint64)
+	for i := range sample {
+		counter[hack.String(sample[i])]++
+	}
+	sorted, onlyOnceItems := make([]uint64, 0, len(counter)), uint64(0)
+	for _, cnt := range counter {
+		sorted = append(sorted, cnt)
+		if cnt == 1 {
+			onlyOnceItems++
+		}
+	}
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i] > sorted[j]
+	})
+
+	var (
+		// last is the last element in top N index should occurres atleast `last` times.
+		last      uint64
+		sumTopN   uint64
+		sampleNDV = uint32(len(sorted))
+	)
+	numTop = mathutil.MinUint32(sampleNDV, numTop) // Ensure numTop no larger than sampNDV.
+	// Only element whose frequency is not smaller than 2/3 multiples the
+	// frequency of the n-th element are added to the TopN statistics. We chose
+	// 2/3 as an empirical value because the average cardinality estimation
+	// error is relatively small compared with 1/2.
+	for i := uint32(0); i < sampleNDV && i < numTop*2; i++ {
+		if i >= numTop && sorted[i]*3 < sorted[numTop-1]*2 && last != sorted[i] {
+			break
+		}
+		last = sorted[i]
+		sumTopN += sorted[i]
+	}
+
+	return &topNHelper{uint64(len(sample)), numTop, counter, sorted, onlyOnceItems, sumTopN, last}
+}
+
+// NewCMSketchWithTopN returns a new CM sketch with TopN elements.
+func NewCMSketchWithTopN(d, w int32, sample [][]byte, numTop uint32, rowCount uint64) *CMSketch {
+	helper := newTopNHelper(sample, numTop)
+	// rowCount is not a accurate value when fast analyzing
+	// In some cases, if user triggers fast analyze when rowCount is close to sampleSize, unexpected bahavior might happen.
+	rowCount = mathutil.MaxUint64(rowCount, uint64(len(sample)))
+	estimateNDV, scaleRatio := calculateEstimateNDV(helper, rowCount)
+	c := buildCMSWithTopN(helper, d, w, scaleRatio)
+	c.calculateDefaultVal(helper, estimateNDV, scaleRatio, rowCount)
+	return c
+}
+
+func buildCMSWithTopN(helper *topNHelper, d, w int32, scaleRatio uint64) (c *CMSketch) {
+	c, helper.sumTopN, helper.numTop = NewCMSketch(d, w), 0, 0
+	enableTopN := helper.sampleSize/topNThreshold <= helper.sumTopN
+	if enableTopN {
+		c.topN = make(map[uint64][]topNMeta)
+	}
+	for counterKey, cnt := range helper.counter {
+		data, scaledCount := hack.Slice(string(counterKey)), cnt*scaleRatio
+		if enableTopN && cnt >= helper.lastVal {
+			h1, h2 := murmur3.Sum128(data)
+			c.topN[h1] = append(c.topN[h1], topNMeta{h1, h2, data, scaledCount})
+			helper.sumTopN += scaledCount
+			helper.numTop++
+		} else {
+			c.updateBytesWithDelta(data, scaledCount)
+		}
+	}
+	return
+}
+
+func (c *CMSketch) calculateDefaultVal(helper *topNHelper, estimateNDV, scaleRatio, rowCount uint64) {
+	sampleNDV := uint64(len(helper.sorted))
+	if rowCount <= helper.sumTopN {
+		c.defaultValue = 1
+	} else if estimateNDV <= uint64(helper.numTop) {
+		c.defaultValue = 1
+	} else if estimateNDV+helper.onlyOnceItems <= uint64(sampleNDV) {
```

Comment:
What is the rationale behind this condition?

Reply:
This conditon is to avoid `estimateNDV - sampleNDV + helper.onlyOnceItems < 0`
However, this will never happen and should be deleted.